File: 0055-jump-game/0055-jump-game.java
class Solution {
    public boolean canJump(int[] nums) {
        int reach = 0;
        int n = nums.length;
        for (int i=0; i<n; i++)
        {
            if (i>reach)
            {
                return false;
            }
            if (i==n-1)
            {
                return true;
            }
            int max = i+nums[i];
            if (max>=n-1)
            {
                return true;
            }
            reach = Math.max(max, reach);
        }
        return false;
    }
}

File: 0055-jump-game/0055-jump-game.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        max_reach = nums[0]
        for i, x in enumerate(nums):
            if max_reach < i:
                return False
            max_reach = max(max_reach, i+x)
            if max_reach >= len(nums)-1:
                return True
        return False
        